### PR TITLE
[Test Only] Add Galaxy sanity test for Gemma4-31B disaggregated prefill

### PR DESCRIPTION
**Pull request by mmanzoorTT**
### PR Category

Test Only

### Summary

Add a single Galaxy sanity test for Gemma4-31B disaggregated prefill and register it in the Blaze Models Prefill pipeline.

The existing D_P coverage is structural and host-side. This change adds one cheap, non-parametrized device leg that runs the canonical production prefill shape end to end on a Blackhole Galaxy 8x4 CP8/TP4 mesh, so regressions that break the path entirely are caught in CI without paying for a full correctness sweep.

### Implemented Test

`models/demos/gemma4_d_p/tests/test_sanity_glx.py::test_sanity_glx`

- Non-parametrized wrapper that calls `text_demo_prefill.py::test_prefill_long_context_traced` at one pinned shape:
  - traced prefill with one replayed ring-attention trace
  - 256K context in 8K chunks
  - final-chunk readback only (`readback_all=False`)
  - real text tokens (`token_source="text"`)
  - 8x4 mesh = CP8/TP4, fabric enabled, `trace_region_size` from `TRACE_REGION_SIZE`
- Equivalent to `test_prefill_long_context_traced[...-readback_final-ctx_256k-chunk8192-text-8x4]`, but reachable as a single `pytest models/demos/gemma4_d_p/tests/test_sanity_glx.py` invocation.
- This is an execution check, not a numerical check: it verifies the D_P traced prefill path still runs end to end on a Galaxy box. Golden trace comparisons are staged separately under `/mnt/models`.

### CI Registration

Register the leg in `tests/pipeline_reorg/blaze_models_prefill_tests.yaml`:

- Name: `(Gemma4-31B-it) D_P prefill sanity 256k@8k`
- Test stage / `test_type`: `gemma4_d_p`
- Tags: `gemma4_d_p`, `transformer`; group `module`; budget type `demo`
- Hardware: Blackhole Galaxy `bh_sc1`, `sp_config: sc1`, `fabric_profile: 1d`
- Mesh: 8x4 CP8/TP4
- Checkpoint: `google/gemma-4-31B-it`
- Timeout: 30 minutes (pytest `--timeout=1800`)
- Tier 1, `release_ready: false`
- Owner: Muhammad Asif Manzoor (`U078RU2KP9Q`)

`gemma4_d_p` is also added to the valid-stages list in `.github/workflows/blaze-models-prefill-tests.yaml` so the stage can be dispatched on its own.

The job runs under `mpirun --bind-to none --pernode` with:

```bash
export HF_HUB_OFFLINE=1
export HF_HOME=/mnt/models/huggingface
export HF_MODEL=google/gemma-4-31B-it
export TT_CACHE_PATH=/mnt/models/huggingface/tt_cache/gemma4_d_p/google--gemma-4-31B-it
```

### Validation

- The pinned 256K@8K traced shape has been exercised on the 8x4 mesh and completes end to end.

### Notes for reviewers

- The test uses the staged offline `google/gemma-4-31B-it` checkpoint and D_P tensor cache under `/mnt/models`; it is not runnable without that staging.
- The wrapper deliberately delegates to the demo test rather than duplicating its setup, so the sanity shape cannot drift from the parametrized coverage.
- 8x4 is the only mesh covered here. The underlying demo test also parametrizes 4x8, but the sanity leg pins the supported production mesh.
- No numerical thresholds are asserted in this PR.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mmanzoor/gemma4-prefill-ci-tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mmanzoor/gemma4-prefill-ci-tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mmanzoor/gemma4-prefill-ci-tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mmanzoor/gemma4-prefill-ci-tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mmanzoor/gemma4-prefill-ci-tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mmanzoor/gemma4-prefill-ci-tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mmanzoor/gemma4-prefill-ci-tests)

### CODE REVIEW

File: tests/pipeline_reorg/blaze_models_prefill_tests.yaml [hunk 1]
```diff
@@ -779,6 +715,38 @@
   sp_config: sc1
 
 
+# ===== Gemma4-31B D/P =================================================================================
+
+# Short numerical gate for the standalone disaggregated-prefill implementation. The test module
+# restricts itself to the supported CP8/TP4 Galaxy layout and exercises both the HF reference
+# comparison and traced-vs-eager replay. The checkpoint and tilized D/P cache are staged on /mnt/models.
+- name: "(Gemma4-31B-it) D/P prefill accuracy 8k"
+  fabric_profile: torus_xy
```
Comment (svuckovicTT) [inline tests/pipeline_reorg/blaze_models_prefill_tests.yaml:724]: I think we're using 1d fabric in the canonical test

File: .github/time_budget.yaml [hunk 2]
```diff
@@ -408,8 +408,8 @@ models:
     bh_quietbox_2: 80
   demo:
     bh_galaxy: 470
-    bh_sc1: 536
-    bh_sc1_high_power: 191
+    bh_sc1: 620
```
Comment (svuckovicTT) [inline .github/time_budget.yaml:411]: Do we need sc1? The whole of prefill runs on a single GLX. That said, we do want a high powered version (14 kW) of glx to run this test, but I don't see a pre-existing option to target that, only sc1.

@dpopovTT @kgrujcic-tt hey guys, could you advise on what we can/should do to get a single high-powered glx for prefill perf tests?

File: .github/workflows/blaze-models-prefill-tests.yaml [hunk 3]
```diff
@@ -12,7 +12,7 @@ on:
...
+          Valid stages: dsa_mla_glm, gemma4_dp, glm_chunked, glm_chunked_accuracy, glm_moe, glm_prefill_block, hca_chunked, hca_perf, high_bw_all_gather_perf, k3_kda, k3_kda_perf, k3_mla, k3_mla_perf, k3_mla_trace, kimi_chunked, kimi_chunked_padded, kimi_dflash, kimi_k3_moe, kimi_moe, kimi_moe_perf, kv_cache_table, minimax_prefill_kv_pcc, minimax_prefill_perf, mistral4_chunked_padded, mistral4_mla, mistral4_mla_chunked, mistral4_moe, mistral4_prefill_block, mistral4_prefill_transformer, mla_chunked, moe_gate, prefill_block_determinism, prefill_runner, prefill_runner_kv_pcc, prefill_transformer_determinism, ring_joint_sdpa_perf.
```
Comment (svuckovicTT) [inline .github/workflows/blaze-models-prefill-tests.yaml:15]: nitpicking here, but I like consistency

Can we rename to `gemma4_d_p` so that it matches the dir path of the model impl?

File: .github/time_budget.yaml (hunk 2 above)
Comment (svuckovicTT) [reply .github/time_budget.yaml:411]: @dpopovTT tells me that sc1 is a single glx:
```
sc1 = single galaxy
sc4 = quad
sc16 = 16 galaxies
...
```

so we're good there. We can use this tag, and then later, for perf tests, we should be able to use the high power tag one line below.

Comment (mmanzoorTT) [reply .github/time_budget.yaml:411]: Thanks for the confirmation.

File: .github/workflows/blaze-models-prefill-tests.yaml (hunk 3 above)
Comment (mmanzoorTT) [reply .github/workflows/blaze-models-prefill-tests.yaml]: updated.

**Review by mtairum — APPROVED** [review]

File: models/demos/gemma4_d_p/tests/test_prefill_correctness.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+        (1024, 1024, [0]),
+        (1024, 2048, [0, 1024]),
+        (1024, 3072, [0, 1024, 2048]),
+        (2048, 4096, [0, 2048]),
+        (2048, 8192, [0, 2048, 4096, 6144]),
+    ],
+)
+def test_chunk_boundary_starts_cover_sliding_window_edges(chunk_size, context_len, expected_starts):
+    starts = [idx * chunk_size for idx in range(context_len // chunk_size)]
+    assert starts == expected_starts
```
Comment (svuckovicTT) [inline models/demos/gemma4_d_p/tests/test_prefill_correctness.py:107]: This test doesn't really test any of our logic, remove.

File: models/demos/gemma4_d_p/tests/test_prefill_correctness.py [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+    ],
+)
+def test_prefill_metadata_tracks_user_and_chunk(chunk_start, user_id):
+    events = []
+    model = _make_model(events)
+    hidden = SimpleNamespace(shape=(1, 1, 128, 64))
+
+    model(hidden_states=hidden, chunk_start_idx=chunk_start, user_id=user_id)
+
+    assert ("ring_metadata", user_id, chunk_start) in events
```
Comment (svuckovicTT) [inline models/demos/gemma4_d_p/tests/test_prefill_correctness.py:92]: This test doesn't isn't of value, I'd remove it.

File: models/demos/gemma4_d_p/tests/test_prefill_correctness.py [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+        (2048, 4096, [0, 2048]),
+        (2048, 8192, [0, 2048, 4096, 6144]),
+    ],
+)
+def test_chunk_boundary_starts_cover_sliding_window_edges(chunk_size, context_len, expected_starts):
+    starts = [idx * chunk_size for idx in range(context_len // chunk_size)]
+    assert starts == expected_starts
+
+
+def test_traced_and_untraced_prefill_dispatch_share_same_chunk_metadata(monkeypatch):
```
Comment (svuckovicTT) [inline models/demos/gemma4_d_p/tests/test_prefill_correctness.py:110]: Doesn't really test the difference between traced and untraced paths, the `_make_model()` call always sets
```
    model._prefill_trace_mode = True
```

The only difference I see is the way rope matrices are supplied.

File: models/demos/gemma4_d_p/tests/test_prefill_correctness.py [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+    del model
+
+
+@torch.no_grad()
+@pytest.mark.timeout(3600)
+@parametrize_mesh_with_fabric([(8, 4)], device_params_extra={"trace_region_size": 256_000_000})
+def test_device_prefill_matches_hf(mesh_device, hf_prefill_reference):
+    """Compare one real 8x4 D/P prefill with the HF final-token logits."""
+    chunk_size = 8192
+    tokens = _get_prefill_tokens(_model_path(), chunk_size, hf_prefill_reference.config.vocab_size, source="random")
```
Comment (svuckovicTT) [inline models/demos/gemma4_d_p/tests/test_prefill_correctness.py:371]: Using random tokens is generally not a good idea when looking to verify accuracy. In the past, we've seen PCC > 0.99 with random tokens, only for it to go down to 0.95 with real tokens.

File: models/demos/gemma4_d_p/demo/text_demo_prefill.py [hunk 8]
```diff
@@ -189,13 +189,14 @@ def _build_prefill_model(mesh_device, model_path, chunk_size, context_len=None):
     t0 = time.time()
     model_args, model, kv_cache, _state_dict = create_tt_model(
         mesh_device=mesh_device,
-        max_batch_size=1,
+        max_batch_size=max_batch_size,
```
Comment (svuckovicTT) [inline models/demos/gemma4_d_p/demo/text_demo_prefill.py:192]: Let's undo the changes in this file - we always want batch size of 1.

**Review by svuckovicTT — CHANGES_REQUESTED** [review]
Hey Asif, thanks for the changes.

Some of these tests have the right idea, but I'd like to see them in a different form. Let's sync offline.

File: tests/pipeline_reorg/blaze_models_prefill_tests.yaml [hunk 9]
```diff
@@ -848,6 +848,39 @@
...
+      python3 -m pytest models/demos/gemma4_d_p/tests/test_sanity_glx.py \
+        -vv --tb=short --timeout=3600
+    '
+  test_type: gemma4_d_p
+  test_tags: [gemma4_d_p, transformer]
+  test_group: module
+  budget_type: demo
+  skus:
+    bh_sc1:
+      timeout: 120
```
Comment (svuckovicTT) [inline tests/pipeline_reorg/blaze_models_prefill_tests.yaml]: This job is failing: https://github.com/tenstorrent/tt-metal/actions/runs/34968456846/job/104378532416?pr=56078

Let's reduce this to 30 min, it should be more than enough. If the kernels are cached, we can go even further. Let's change it to 30 and run the model test to verify before merging.

Comment (svuckovicTT) [reply tests/pipeline_reorg/blaze_models_prefill_tests.yaml]: Was curious how this works so started a run here: https://github.com/tenstorrent/tt-metal/actions/runs/34973115034

Comment (mmanzoorTT) [reply tests/pipeline_reorg/blaze_models_prefill_tests.yaml]: @svuckovicTT This job failed due to `precision override` issue; fixed in https://github.com/tenstorrent/tt-metal/pull/56636

Re-ran the Gemma4 blaze test here: https://github.com/tenstorrent/tt-metal/actions/runs/35058924745. Successfully ran the Gemma4 test but failed to upload the log report due to invalid character in test name. Fixed in `fdc58612ff8`

**Review by mmanzoorTT — COMMENTED** [review]
fdc58612ff8

**Review by svuckovicTT — APPROVED** [review]

**Review by dpopovTT — APPROVED** [review]
